File: packages/kits/CorpusKit/Tests/CorpusKitTests/DefaultEnsembleRecallPayoffTests.swift
// DefaultEnsembleRecallPayoffTests.swift
//
// Mission 6a-iii-wire — end-to-end PAYOFF proof: the default recall ensemble
// (CorpusEnsemble.defaultEnsemble(): RI/PPMI/LSA/NMF/FDC) un-pins recall.
//
// This is the deterministic, CI-verifiable proof that flipping the production
// default from a single hash lane to the five honest signals actually fixes
// recall. The Corpus built here is exactly the one GLK's `provision` default and
// the ARIA_MCP server now construct (both thread `CorpusEnsemble.defaultEnsemble()`
// into `Corpus(storage:models:)`), so proving the un-pinning at this layer proves
// it for every production provision site.
//
// What "un-pinning" means and why it is the payoff
// ------------------------------------------------
// A single fake/hash lane (the old `.deterministic` default) collapses recall:
// its similarity reflects only surface byte overlap, so varied queries tend to
// pin onto the same handful of lexically-overlapping documents, and a query that
// is semantically related but lexically different misses entirely. The five-signal
// ensemble — trained on the estate's own corpus (RI/PPMI/LSA/NMF) plus stateless
// taxonomic FDC — produces distributional + categorical structure, so:
//   (a) varied queries return DIVERSE top hits (not pinned to one cluster),
//   (b) every hit carries MULTI-SIGNAL dense provenance (multiple modelIDs vote),
//   (c) a semantically-related-but-lexically-different document is recalled.
//
// Determinism: `now` is a fixed Date; all five providers are deterministic
// (fixed seeds / fixed-iteration training). No Date() in the engine path.
//
// Test isolation: Corpus ingest/reindex emit corpuskit.* metrics through the
// global Intellectus sink; the telemetry suite asserts an exact count under a
// captured window, so every Corpus-op suite serialises via GlobalTestLock. This
// suite follows that convention.

import Testing
import Foundation
import CorpusKit
import CorpusKitProviders
import PersistenceKit
import PersistenceKitSQLite

@Suite("DefaultEnsemble recall payoff", .serialized)
struct DefaultEnsembleRecallPayoffTests {

    private let now = Date(timeIntervalSince1970: 1_700_000_000)

    @Test("Default ensemble invalidates trainable 1.0 bases")
    func defaultEnsembleModelVersions() {
        let versions = CorpusEnsemble.defaultEnsemble().compactMap { model -> String? in
            switch model {
            case .randomIndexing(let provider), .ppmi(let provider),
                 .lsa(let provider), .nmf(let provider), .fdc(let provider):
                return provider.modelVersion
            default:
                return nil
            }
        }
        #expect(versions == ["1.1.0", "1.1.0", "1.1.0", "1.1.0", "1.0.0"])
    }

    /// A diverse multi-topic corpus spanning four clearly separated topical
    /// clusters (space / cooking / finance / gardening). Enough distinct topics
    /// that a single surface-hash lane would collapse recall onto a tiny set,
    /// while the trained distributional signals resolve the clusters. Each doc is
    /// one chunk whose text equals the doc.
    private let docs: [(id: String, text: String)] = [
        // space cluster
        ("space-1", "rocket launch orbit satellite spacecraft mission"),
        ("space-2", "astronaut spacecraft orbit station module docking"),
        ("space-3", "telescope galaxy star planet nebula cosmos observation"),
        // cooking cluster
        ("cook-1", "recipe oven bake bread flour yeast dough"),
        ("cook-2", "saute pan onion garlic simmer sauce stove"),
        ("cook-3", "knife chop vegetable dice prep cutting board"),
        // finance cluster
        ("fin-1", "invest portfolio stock bond dividend market return"),
        ("fin-2", "budget savings expense income loan interest rate"),
        ("fin-3", "tax filing deduction revenue accounting ledger audit"),
        // gardening cluster
        ("garden-1", "soil seed plant water sunlight grow sprout"),
        ("garden-2", "prune shrub hedge trim branch leaf foliage"),
        ("garden-3", "compost fertilizer nutrient root mulch garden bed")
    ]

    private func scratchURL() -> URL {
        FileManager.default.temporaryDirectory
            .appendingPathComponent("corpuskit-payoff-\(UUID().uuidString).sqlite3")
    }

    private func storage(at url: URL) throws -> any Storage {
        try SQLiteStorage(configuration: EstateConfiguration(
            estateID: UUID(),
            backend: .sqlite(url: url, busyTimeout: 5.0)))
    }

    /// Build a Corpus on the canonical default ensemble, ingest the diverse
    /// corpus, and reindex (which trains all four trainable signals on the full
    /// corpus). Returns the trained-ready Corpus.
    private func makeTrainedEnsembleCorpus() async throws -> Corpus {
        // THE call under test: the exact default GLK.provision and the ARIA_MCP
        // server now thread. If this set ever drifts from the production default,
        // the payoff proof drifts with it.
        let corpus = try await Corpus(
            storage: try storage(at: scratchURL()),
            models: CorpusEnsemble.defaultEnsemble())
        for doc in docs {
            try await corpus.ingest(doc.text, sourceID: doc.id, now: now)
        }
        // reindex trains the four trainable signals (RI/PPMI/LSA/NMF) on the full
        // corpus and re-embeds every chunk under every signal's modelID. FDC is
        // stateless and needs no training.
        try await corpus.reindex(now: now)
        return corpus
    }

    /// Pull ranked itemIDs out of a per-signal result for the default (slot-0)
    /// signal — the signal the single-signal `floatNearest`/`recall` delegate to.
    private func rankedIDs(_ outcome: FloatLaneOutcome) -> [String] {
        if case .hits(let pairs) = outcome { return pairs.map(\.itemID) }
        return []
    }

    // MARK: - (a) varied queries return DIVERSE hits (recall is NOT pinned)

    @Test("Varied queries recall diverse documents, not one pinned cluster")
    func variedQueriesAreNotPinned() async throws {
        try await GlobalTestLock.shared.withLock {
            let corpus = try await makeTrainedEnsembleCorpus()

            // Four queries, one per topical cluster. Each query's words appear in
            // its cluster's docs, so a working recall returns that cluster's docs
            // near the top — and crucially the TOP hit differs across queries.
            let queries: [(probe: String, expectCluster: String)] = [
                ("orbit spacecraft mission", "space"),
                ("bake bread oven", "cook"),
                ("invest stock portfolio", "fin"),
                ("plant soil water grow", "garden")
            ]

            var topHits: [String] = []
            for q in queries {
                let outcome = await corpus.floatNearest(query: q.probe, limit: 3)
                let ids = rankedIDs(outcome)
                #expect(!ids.isEmpty, "query '\(q.probe)' must return ranked hits")
                let top = try #require(ids.first)
                topHits.append(top)
                // The top hit must belong to the query's own cluster — proof the
                // ensemble resolves topic structure, not surface noise.
                #expect(
                    top.hasPrefix(q.expectCluster),
                    "query '\(q.probe)' top hit '\(top)' must be in cluster '\(q.expectCluster)'")
            }

            // UN-PINNING: the four queries do NOT collapse onto the same handful
            // of documents. A pinned (single-hash) lane would repeat the same ids;
            // the trained ensemble returns four DISTINCT top hits.
            #expect(
                Set(topHits).count == queries.count,
                "varied queries must recall DISTINCT top documents (un-pinned), got \(topHits)")
        }
    }

    // MARK: - (b) hits carry MULTI-SIGNAL dense provenance (multiple modelIDs vote)

    @Test("Recall hits carry multi-signal dense provenance (>1 modelID voting)")
    func hitsCarryMultiSignalProvenance() async throws {
        try await GlobalTestLock.shared.withLock {
            let corpus = try await makeTrainedEnsembleCorpus()

            let perSignal = await corpus.floatNearestPerSignal(
                query: "orbit spacecraft mission", limit: 3)

            // The default ensemble holds five signals; each appears once, in slot
            // order, tagged by its own modelID.
            let modelIDs = perSignal.map(\.modelID)
            #expect(
                modelIDs == ["random-indexing-v1", "ppmi-v1", "lsa-v1", "nmf-v1", "fdc-v1"],
                "per-signal provenance must carry all five default modelIDs in order, got \(modelIDs)")

            // MULTI-SIGNAL VOTING: more than one signal must produce ranked hits
            // for the query (the single-hash default could only ever produce one
            // lane). Count signals whose lane returned ranked hits.
            let votingSignals = perSignal.filter { !rankedIDs($0.outcome).isEmpty }
            #expect(
                votingSignals.count >= 2,
                "at least two dense signals must vote on the query, got \(votingSignals.count) (\(votingSignals.map(\.modelID)))")

            // And the trained distributional signals (the heart of the honest
            // semantic lane) must agree the top hit is in the space cluster.
            for signal in votingSignals where signal.modelID != "fdc-v1" {
                let top = rankedIDs(signal.outcome).first
                #expect(
                    top?.hasPrefix("space") == true,
                    "signal \(signal.modelID) top hit \(top ?? "nil") should be in space cluster")
            }
        }
    }

    // MARK: - (c) semantically-related-but-lexically-different recall (BM25 misses)

    @Test("Ensemble recalls a semantically-related, lexically-different document")
    func semanticNotLexicalRecall() async throws {
        try await GlobalTestLock.shared.withLock {
            let corpus = try await makeTrainedEnsembleCorpus()

            // The probe uses ONLY cook-1's baking vocabulary ("oven bake flour
            // dough"). The target document cook-3 ("knife chop vegetable dice prep
            // cutting board") shares ZERO tokens with the probe — a pure lexical
            // (BM25) match scores cook-3 at zero and never surfaces it. But the
            // trained distributional ensemble learned that cook-1, cook-2, and
            // cook-3 co-occur in the same cooking neighbourhood, so the dense lane
            // places cook-3 near the baking probe even with no shared surface token.
            let probeTokens: Set<String> = ["oven", "bake", "flour", "dough"]
            let cook3Tokens: Set<String> = ["knife", "chop", "vegetable", "dice", "prep", "cutting", "board"]
            // Guard the premise: probe and target genuinely share no token, so a
            // lexical match could not surface cook-3.
            #expect(probeTokens.isDisjoint(with: cook3Tokens),
                    "test premise: probe and cook-3 must share no surface token")

            let outcome = await corpus.floatNearest(query: "oven bake flour dough", limit: 12)
            let ids = rankedIDs(outcome)
            #expect(!ids.isEmpty, "semantic probe must return ranked hits")

            // cook-3 — semantically related, lexically disjoint — must be recalled.
            // This is precisely the hit BM25 alone misses.
            #expect(
                ids.contains("cook-3"),
                "ensemble must recall the lexically-disjoint cooking doc cook-3; recalled: \(ids)")
        }
    }
}

File: packages/kits/CorpusKit/rust/tests/default_ensemble_recall_payoff_tests.rs
//! Mission 6a-iii-wire — end-to-end PAYOFF proof (Rust parity leg).
//!
//! The default recall ensemble (`corpus_kit_providers::default_ensemble()`:
//! RI then LSA) un-pins recall. This is the Rust mirror of the Swift
//! `DefaultEnsembleRecallPayoffTests`: the Corpus built here is exactly the one
//! the Rust GLK provision path (via the moot-mgr / ARIA_MCP app callers) and
//! `estate_registry` now construct (all thread `default_ensemble()` into
//! `Corpus::open_many`), so proving the un-pinning here proves it for every Rust
//! production provision site.
//!
//! What "un-pinning" means and why it is the payoff
//! ------------------------------------------------
//! A single fake/hash lane (the old `Deterministic` default) collapses recall
//! onto a handful of lexically-overlapping documents and misses
//! semantically-related-but-lexically-different content. The two-signal
//! ensemble — both signals trained on the estate's own corpus — produces
//! distributional structure, so:
//!   (a) varied queries return DIVERSE top hits (not pinned to one cluster),
//!   (b) every hit carries MULTI-SIGNAL dense provenance (multiple model_ids vote),
//!   (c) a semantically-related-but-lexically-different document is recalled.
//!
//! Real SQLite (file-backed), never InMemory: the same primitive-form read-back
//! discipline as the other corpus integration tests.

// The payoff under proof is the default dense ensemble: RI and LSA, both always-on.

use corpus_kit::{Corpus, EmbeddingModelConfig, FloatLaneOutcome};
use corpus_kit_providers::default_ensemble;
use persistence_kit::{BackendConfiguration, EstateConfiguration, SqliteStorage, Storage};
use std::collections::HashSet;
use std::sync::{Arc, Mutex, OnceLock};
use uuid::Uuid;

// Process-wide lock: Corpus.ingest / reindex emit IntellectusLib telemetry and
// the SQLite scratch files must not race. Shared discipline with corpus_tests.
static GLOBAL_LOCK: OnceLock<Mutex<()>> = OnceLock::new();
fn global_lock() -> std::sync::MutexGuard<'static, ()> {
    match GLOBAL_LOCK.get_or_init(|| Mutex::new(())).lock() {
        Ok(g) => g,
        Err(p) => p.into_inner(),
    }
}

/// A diverse multi-topic corpus spanning four clearly separated topical clusters
/// (space / cooking / finance / gardening), identical to the Swift
/// `DefaultEnsembleRecallPayoffTests.docs`. Each doc is one chunk.
const DOCS: [(&str, &str); 12] = [
    ("space-1", "rocket launch orbit satellite spacecraft mission"),
    ("space-2", "astronaut spacecraft orbit station module docking"),
    ("space-3", "telescope galaxy star planet nebula cosmos observation"),
    ("cook-1", "recipe oven bake bread flour yeast dough"),
    ("cook-2", "saute pan onion garlic simmer sauce stove"),
    ("cook-3", "knife chop vegetable dice prep cutting board"),
    ("fin-1", "invest portfolio stock bond dividend market return"),
    ("fin-2", "budget savings expense income loan interest rate"),
    ("fin-3", "tax filing deduction revenue accounting ledger audit"),
    ("garden-1", "soil seed plant water sunlight grow sprout"),
    ("garden-2", "prune shrub hedge trim branch leaf foliage"),
    ("garden-3", "compost fertilizer nutrient root mulch garden bed"),
];

const NOW_MILLIS: i64 = 1_700_000_000_000;

/// Both default signals carry the 1.1.0 basis version, so a persisted 1.0
/// basis is invalidated and retrained rather than restored.
#[test]
fn default_ensemble_invalidates_trainable_1_0_bases() {
    let models = default_ensemble();
    let versions: Vec<&str> = models
        .iter()
        .map(|model| match model {
            EmbeddingModelConfig::RandomIndexing { provider }
            | EmbeddingModelConfig::Lsa { provider } => provider.model_version(),
            _ => panic!("unexpected model in the default ensemble"),
        })
        .collect();
    assert_eq!(versions, ["1.1.0", "1.1.0"]);
}

fn scratch_path() -> String {
    std::env::temp_dir()
        .join(format!("corpuskit-payoff-rust-{}.sqlite3", Uuid::new_v4()))
        .to_string_lossy()
        .into_owned()
}

fn storage_at(path: &str) -> Arc<dyn Storage> {
    let config = EstateConfiguration::new(
        Uuid::new_v4(),
        BackendConfiguration::Sqlite {
            path: path.to_string(),
            busy_timeout_secs: 5.0,
        },
    );
    Arc::new(SqliteStorage::new(config).expect("open sqlite"))
}

/// Build a Corpus on the canonical default ensemble, ingest the diverse
/// corpus, and reindex (trains both signals). The call under test:
/// `default_ensemble()` — the exact set the Rust production provision sites thread.
fn make_trained_ensemble_corpus() -> Corpus {
    let corpus =
        Corpus::open_many(storage_at(&scratch_path()), default_ensemble()).expect("open_many");
    for (id, text) in DOCS.iter() {
        corpus.ingest(text, id, NOW_MILLIS).expect("ingest");
    }
    corpus.reindex(NOW_MILLIS).expect("reindex");
    corpus
}

/// Pull ranked item ids out of a FloatLaneOutcome (default / per-signal).
fn ranked_ids(outcome: &FloatLaneOutcome) -> Vec<String> {
    match outcome {
        FloatLaneOutcome::Hits(pairs) => pairs.iter().map(|(id, _)| id.clone()).collect(),
        _ => Vec::new(),
    }
}

// (a) varied queries return DIVERSE hits (recall is NOT pinned)
#[test]
fn varied_queries_are_not_pinned() {
    let _guard = global_lock();
    let corpus = make_trained_ensemble_corpus();

    // Four queries, one per topical cluster; the top hit must differ across them.
    let queries: [(&str, &str); 4] = [
        ("orbit spacecraft mission", "space"),
        ("bake bread oven", "cook"),
        ("invest stock portfolio", "fin"),
        ("plant soil water grow", "garden"),
    ];

    let mut top_hits: Vec<String> = Vec::new();
    for (probe, cluster) in queries.iter() {
        let outcome = corpus.float_nearest(probe, 3);
        let ids = ranked_ids(&outcome);
        assert!(!ids.is_empty(), "query '{probe}' must return ranked hits");
        let top = ids[0].clone();
        assert!(
            top.starts_with(cluster),
            "query '{probe}' top hit '{top}' must be in cluster '{cluster}'"
        );
        top_hits.push(top);
    }

    // UN-PINNING: the four queries do NOT collapse onto the same documents.
    let distinct: HashSet<&String> = top_hits.iter().collect();
    assert_eq!(
        distinct.len(),
        queries.len(),
        "varied queries must recall DISTINCT top documents (un-pinned), got {top_hits:?}"
    );
}

// (b) hits carry MULTI-SIGNAL dense provenance (multiple model_ids vote)
#[test]
fn hits_carry_multi_signal_provenance() {
    let _guard = global_lock();
    let corpus = make_trained_ensemble_corpus();

    let per_signal = corpus.float_nearest_per_signal("orbit spacecraft mission", 3);

    let model_ids: Vec<&str> = per_signal.iter().map(|(id, _)| id.as_str()).collect();
    assert_eq!(
        model_ids,
        vec!["random-indexing-v1", "lsa-v1"],
        "per-signal provenance must carry both default model_ids in order, got {model_ids:?}"
    );

    // MULTI-SIGNAL VOTING: more than one signal must produce ranked hits.
    let voting: Vec<&str> = per_signal
        .iter()
        .filter(|(_, o)| !ranked_ids(o).is_empty())
        .map(|(id, _)| id.as_str())
        .collect();
    assert!(
        voting.len() >= 2,
        "at least two dense signals must vote on the query, got {} ({voting:?})",
        voting.len()
    );

    // Both trained distributional signals must agree the top hit is space-cluster.
    for (model_id, outcome) in per_signal.iter() {
        let ids = ranked_ids(outcome);
        if let Some(top) = ids.first() {
            assert!(
                top.starts_with("space"),
                "signal {model_id} top hit {top} should be in space cluster"
            );
        }
    }
}

// (c) semantically-related-but-lexically-different recall (BM25 misses)
#[test]
fn semantic_not_lexical_recall() {
    let _guard = global_lock();
    let corpus = make_trained_ensemble_corpus();

    // Probe = cook-1's baking vocabulary only. cook-3 shares ZERO tokens with the
    // probe, so a lexical (BM25) match scores it at zero. The trained ensemble
    // learned cook-1/2/3 co-occur, so the dense lane still surfaces cook-3.
    let probe_tokens: HashSet<&str> = ["oven", "bake", "flour", "dough"].into_iter().collect();
    let cook3_tokens: HashSet<&str> =
        ["knife", "chop", "vegetable", "dice", "prep", "cutting", "board"]
            .into_iter()
            .collect();
    assert!(
        probe_tokens.is_disjoint(&cook3_tokens),
        "test premise: probe and cook-3 must share no surface token"
    );

    let outcome = corpus.float_nearest("oven bake flour dough", 12);
    let ids = ranked_ids(&outcome);
    assert!(!ids.is_empty(), "semantic probe must return ranked hits");
    assert!(
        ids.iter().any(|id| id == "cook-3"),
        "ensemble must recall the lexically-disjoint cooking doc cook-3; recalled: {ids:?}"
    );
}

// MARK: - Gate proofs: default ensemble composition and LSA lane

/// Gate proof 1 — `default_ensemble()` must return exactly two configs: RI then LSA.
/// Validates the always-on ensemble contract (DENSE_LANE_TRIM ruling).
#[test]
fn default_ensemble_is_ri_and_lsa() {
    let ensemble = default_ensemble();
    assert_eq!(ensemble.len(), 2, "default ensemble must contain exactly two signals");
    assert!(
        matches!(ensemble[0], EmbeddingModelConfig::RandomIndexing { .. }),
        "first config must be RandomIndexing (RI)"
    );
    assert!(
        matches!(ensemble[1], EmbeddingModelConfig::Lsa { .. }),
        "second config must be Lsa"
    );
}

/// Gate proof 2 — the LSA lane encodes two drawers with disjoint vocabulary
/// and ranks the on-topic one first. Uses the public
/// `Corpus::float_nearest_per_signal` surface on a real SQLite scratch estate
/// and reads the `lsa-v1` entry, so the proof is about the LSA signal itself
/// and not the default (RI) slot that `float_nearest` serves.
#[test]
fn lsa_lane_float_nearest_returns_hits_on_scratch_estate() {
    let _guard = global_lock();
    let storage = storage_at(&scratch_path());
    let corpus = Corpus::open_many(storage, default_ensemble())
        .expect("Corpus::open_many must succeed with default ensemble");

    corpus.ingest("rocket launch orbit satellite spacecraft mission", "doc-space", NOW_MILLIS)
        .expect("ingest doc-space");
    corpus.ingest("recipe oven bake bread flour yeast dough", "doc-cook", NOW_MILLIS)
        .expect("ingest doc-cook");
    // Reindex trains the RI and LSA bases; untrained distributional lanes hold
    // no float rows. The query is drawn from doc-space's vocabulary so the
    // trained basis folds it in (an out-of-vocabulary query has no vector).
    corpus.reindex(NOW_MILLIS).expect("reindex");

    let per_signal = corpus.float_nearest_per_signal("rocket orbit", 10);
    let (_, outcome) = per_signal
        .iter()
        .find(|(model_id, _)| model_id == "lsa-v1")
        .expect("the default ensemble holds an lsa-v1 signal");
    match outcome {
        FloatLaneOutcome::Hits(hits) => {
            let ids: Vec<&str> = hits.iter().map(|(id, _)| id.as_str()).collect();
            assert_eq!(
                ids.first().copied(),
                Some("doc-space"),
                "the LSA lane must rank the on-topic drawer first, got {ids:?}"
            );
            assert_ne!(
                ids.last().copied(),
                Some("doc-space"),
                "doc-cook must rank below doc-space when both are returned, got {ids:?}"
            );
        }
        other => panic!("expected Hits from the LSA lane on the scratch estate, got {other:?}"),
    }
}
